proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,700 @@
+            },
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that a `peer` was found for the given `urn`.
+    ///
+    /// If the underlying `Request` was in the `IsRequested` state then it will transition to the
+    /// `Found` state.
+    ///
+    /// If the underlying `Request` was in the `Found` or `Cloning` state then it add this `peer`
+    /// to the set of found peers.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn found(&mut self, urn: &RadUrn, peer: PeerId, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        self.transition(
+            |request| match request {
+                SomeRequest::Requested(request) => Some(request.first_peer(peer, timestamp).into()),
+                SomeRequest::Found(request) => {
+                    let some_request: SomeRequest<T> = request.found(peer, timestamp).into();
+                    Some(some_request)
+                },
+                SomeRequest::Cloning(request) => {
+                    let some_request: SomeRequest<T> = request.found(peer, timestamp).into();
+                    Some(some_request)
+                },
+                _ => None,
+            },
+            Ok,
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that we are attempting a clone from the `peer` for the given `urn`.
+    ///
+    /// If the underlying `Request` was in the `Found` state then it will transition to the
+    /// `Cloning` state.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    ///   * If the underlying `Request` timed out.
+    pub fn cloning(&mut self, urn: &RadUrn, peer: PeerId, timestamp: T) -> Result<(), Error>
+    where
+        T: Clone,
+    {
+        let max_queries = self.config.max_queries;
+        let max_clones = self.config.max_clones;
+        self.transition(
+            |request| match request {
+                SomeRequest::Found(request) => Some(request),
+                _ => None,
+            },
+            |previous| match previous.cloning(max_queries, max_clones, peer, timestamp) {
+                Either::Left(timeout) => Err(timeout.into()),
+                Either::Right(request) => Ok(request),
+            },
+            urn,
+        )
+    }
+
+    /// Tell the `WaitingRoom` that we failed the attempt to clone from the `peer` for the given
+    /// `urn`.
+    ///
+    /// If the underlying `Request` was in the `Found` state then it will transition to the
+    /// `Cloning` state.
+    ///
+    /// # Errors
+    ///
+    ///   * If the `urn` was not in the `WaitingRoom`.
+    ///   * If the underlying `Request` was not in the expected state.
+    pub fn failed(&mut self, peer: PeerId, urn: &RadUrn, timestamp: T) -> Result<(), Error>
```

Comment:
```suggestion
    pub fn failed(&mut self, peer: PeerId, urn: &RadUrn, timestamp: T) -> Result<(), Error>
```

Might be a bit ambiguous otherwise.

Reply:
I'm playing spot the difference here and I'm losing

Reply:
Ah, haha thanks for the edit :+1: